### PR TITLE
Add router-level code splitting using react-loadable

### PR DESCRIPTION
**Pull request by zameschua**
Working on #398 #319

Slightly improves lighthouse report
Lighthouse report can be further improved by following the PRPL pattern

Before:
image
image

After:
image
image

### CODE REVIEW

File: v3/package.json [hunk 1]
```diff
@@ -76,6 +76,7 @@
   },
   "dependencies": {
     "axios": "^0.16.2",
+    "babel-plugin-syntax-dynamic-import": "^6.18.0",
```
Comment (yangshun) [inline v3/package.json]: Should be in `devDependencies`.

File: v3/src/js/views/Loader.jsx [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// @flow
+import React from 'react';
+
+export default function Loader() {
+  return (
+    <div className="col-12 text-center">
```
Comment (yangshun) [inline v3/src/js/views/Loader.jsx]: `col` classes must always be wrapped in a `row`, hence it's not a good idea to leave it the parent to ensure that they have a `.row` container. Should add it here. Then again, you don't need a `.row` and `.col` if you're just doing `col-12`. Simple `div` would suffice.

File: v3/flow-typed/package-dep-libdefs.js [hunk 3]
```diff
@@ -1,9 +1,17 @@
 // This file is autogenerated by flow-scripts.
 
+declare module 'babel-plugin-syntax-dynamic-import' {
```
Comment (yangshun) [inline v3/flow-typed/package-dep-libdefs.js]: After you remove `babel-plugin-syntax-dynamic-import` from `dependencies`, and regenerate this file, this line shouldn't be around.

**Comment by ZhangYiJiang** [comment]
I'm not sure if this is a good idea. This change doesn't seem like it would significantly improve loading time. 

Before: 

```
  203.23 KB     build/vendor.773dbe967b0dc197a3b5.js
  36.52 KB      build/app.2664b4abcacf976da80e.js
```

After: 

```
  203.31 KB  build/vendor.353a389311bffb84d45f.js
  19.79 KB   build/app.2413c47a1a60fcbb164b.js
  10.74 KB   build/3ff70d6b3349cc40471c.js
  6.6 KB     build/3fa02c33b490d2c0c301.js
  1.97 KB    build/0ece4f1a18e432c95372.js
  1.95 KB    build/7c8b75a172bd34d84758.js
  1.79 KB    build/1b4b954b236da1735b8d.js
  1.58 KB    build/068d14c40c31a34be353.js
  1.18 KB    build/8481ed07e8913d1f2c9e.js
  817 B      build/97851965c97bae9fc66e.js
  482 B      build/17a6b53c45fad545ae63.js
```

(CSS and manifest omitted for clarity)

Notice all we've done is split the 36.5 kb app.js down to a bunch of smaller files. The problem is as a percentage of total assets, `app.js` is relatively small. Trimming it further would likely not improve load time, but instead annoy the user as there's now a spinner whenever a new page is loaded for he first time. 

A better option would be to look at `vendor.js`. If you use source-map-explorer, you can see that d3.js takes up a huge amount of space. 

screenshot from 2017-08-11 02-11-59

Since d3 is only used for module details page (dependency graph), selectively loading it *would* make a big impact. (It's probably also a good idea to try to import d3 modules more selectively to allow unused parts of the library to be shaken away, but that's for another time).

**Comment by yangshun** [comment]
I think with code splitting we can exclude some stuff out of vendor bundle. @ZhangYiJiang if you have time would you mind trying that out and see if we can reduce the size of vendor bundle and lazy load d3?

**Comment by ZhangYiJiang** [comment]
Yes, definitely. Sorry my comment wasn't suppose to imply that code splitting as a whole is useless, just that the current approach of splitting the pages is likely to be ineffective. I've actually done vendor library splitting for 3216 before (for Pinpoint.me, the AR portion didn't work on iOS, so we only loaded it on demand for Android users, shaving about 1MB off the original bundle size) - this should be similar.

**Comment by yangshun** [comment]
> Sorry my comment wasn't suppose to imply that code splitting as a whole is useless, just that the current approach of splitting the pages is likely to be ineffective.

No problems, I understand. The groundwork @zameschua has done will nevertheless be useful for enabling the better code splitting approach later on. I personally haven't delved into this much and the project would benefit from your experience (;

File: v3/.eslintrc.js [hunk 4]
```diff
@@ -1,5 +1,9 @@
 module.exports = {
   parser: 'babel-eslint',
+  parserOptions: {
```
Comment (li-kai) [inline v3/.eslintrc.js]: I tried in out locally and removing these additions don't affect anything. Looking through stackoverflow it seems this was used to fix eslint at one point of time.

**Comment by li-kai** [comment]
Nice work. All's left is `git rebase -i upstream/master` (replace upstream with what you named our repo). Replace all but the oldest one with `fixup`, so you are only left with one commit. Remember to remove `/dll` too.

**Comment by zameschua** [comment]
@li-kai Sorry I thought rebasing would be an automatic thing HAHAHA ok on it

**Comment by zameschua** [comment]
**Before
  203.21 KB  build\vendor.82eb0beb7d45e3c82e61.js <-- Contains all vendors including d3.js
  32.98 KB   build\app.6528e664b0f75fcd8393.js
  6.09 KB    build\d314bc55a85689d4a07a.js <-- Contains ModulePageContainer

**After
  127.15 KB  build\vendor.99ffcd2921460b9cbd0c.js <-- Contains all vendor bundles excluding d3.js
  81.96 KB   build\efda2075a88e39938049.js <-- Contains d3.js and ModulePageContainer
  32.97 KB   build\app.3fa6788e70de21d0d011.js

(CSS and Manifest omitted for clarity)

**Comment by yangshun** [comment]
This is great Zames! Perfect! (:

**Pull request merged**